src/daemon/pcsc_context.cpp

```diff
@@ -0,0 +1,379 @@
+                    g_warning("SCardDisconnect() failed: %s", pcsc_stringify_error(ret));
+            }
+
+            std::size_t transmit(const std::vector<std::uint8_t> &send_buffer,
+                                 std::vector<std::uint8_t> &recv_buffer)
+            {
+                DWORD recv_length = recv_buffer.size();
+
+                LONG ret = SCardTransmit(*handle_,
+                                         send_pci_,
+                                         send_buffer.data(),
+                                         send_buffer.size(),
+                                         nullptr,
+                                         recv_buffer.data(),
+                                         &recv_length);
+
+                if (ret != SCARD_S_SUCCESS) {
+                    g_warning("SCardTransmit() failed: %s", pcsc_stringify_error(ret));
+                    return 0;
+                }
+
+                return recv_length;
+            }
+
+        private:
+            std::optional<SCARDHANDLE> handle_;
+            const SCARD_IO_REQUEST *send_pci_ = nullptr;
+        };
+
+        constexpr char NOTIFICATION_READER_NAME[] = R"(\\?PnP?\Notification)";
+        constexpr unsigned int NOTIFICATION_STATE_INDEX = 0;
+
+        std::vector<std::string> multi_string_split(const char *multi_string)
+        {
+            std::vector<std::string> strings;
+            const char *pos = multi_string;
+
+            while (*pos) {
+                strings.emplace_back(pos);
+                pos += strings.back().size() + 1;
+            }
+
+            return strings;
+        }
+
+        std::vector<std::string> list_readers(SCARDCONTEXT context)
+        {
+            // Querying buffer size and then filling buffer with two calls to SCardListReaders() is
+            // racey in pcsc-lite. Each call updates reader list from service. Will most likely fail
+            // if first call reports X readers and next call reports X+1 readers due to buffer being
+            // too small. The X+1:th reader will never be noticed. Have seen this happen so use
+            // SCARD_AUTOALLOCATE even though it requires ugly casting of buffer argument and is not
+            // exception safe (if multi_string_split() OOM:s we want to terminate anyway).
+            char *readers_multi_string;
+            auto readers_multi_string_size = static_cast<DWORD>(SCARD_AUTOALLOCATE);
+            LONG ret;
+
+            ret = SCardListReaders(context,
+                                   nullptr,
+                                   // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                                   reinterpret_cast<LPSTR>(&readers_multi_string),
+                                   &readers_multi_string_size);
+
+            if (ret != SCARD_S_SUCCESS) {
+                if (ret != SCARD_E_NO_READERS_AVAILABLE)
+                    g_warning("Failed to list PC/SC readers: %s", pcsc_stringify_error(ret));
```

Comment:
would prefer to put g_warning into a {}-block.

General comment, has seen seen at other places too.

Reference: AUTOSAR C++14, Rule M6-4-1 (required, implementation, automated)
An if ( condition ) construct shall be followed by a compound statement. The
else keyword shall be followed by either a compound statement, or another
if statement.

Reply:
This can be enforced by https://clang.llvm.org/extra/clang-tidy/checks/readability-braces-around-statements.html . Thresold is currently set to 4 lines (if a statment is split into more lines than that, clang-tidy will complain).

Removing setting of `ShortStatementLines` in `.clang-tidy` will add braces after if/while/do. Will do that on new code added in this pr and open a story for changing it permanently later (in all repos).

Reply:
(Personal preference is to not add `{` and `}` when it clutters the code. Warnings/Errors are available in compilers and `clang-tidy` for mis-indentation, multi-line macros etc. that this convention is meant to protect against. But AUTOSAR C++14 is clear in this case. :)

Reply:
Rule M6-3-1 mentions more statements than `if`.

Reply:
Pushed a fixup commit that makes new code follow AUTOSAR C++14 rule.